packages/backend/src/open_payments/auth/middleware.ts

```diff
@@ -0,0 +1,134 @@
+import assert from 'assert'
+import axios from 'axios'
+
+import {
+  AccessType,
+  AccessAction,
+  Grant,
+  GrantOptions,
+  GrantAccess,
+  GrantAccessJSON
+} from './grant'
+import { AppContext } from '../../app'
+
+export function createAuthMiddleware({
+  type,
+  action
+}: {
+  type: AccessType
+  action: AccessAction
+}) {
+  return async (
+    ctx: AppContext,
+    next: () => Promise<unknown>
+  ): Promise<void> => {
+    try {
+      if (
+        !ctx.request.headers.authorization ||
+        ctx.request.headers.authorization.split(' ')[0] !== 'GNAP'
+      ) {
+        ctx.throw(401, 'Unauthorized')
+      }
+      const token = ctx.request.headers.authorization.split(' ')[1]
```

Comment:
I'm not sure why typescript isn't catching it, but I think `token` can be `undefined` here (when `ctx.request.headers.authorization === 'GNAP'`).